src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -405,6 +614,139 @@ internal void WriteNestedResourceInfo(EntityDescriptor entityDescriptor, IEnumer
+                clientType = model.GetClientTypeAnnotation(model.GetOrCreateEdmType(entityDescriptor.Entity.GetType()));
+            }
+
+            
+
+            foreach (KeyValuePair<string, List<Descriptor>> relatedLinks in groupRelatedLinks)
+            {
+                bool isCollection = false;
+                bool started = false;
+
+                LinkDescriptor relatedEnd = null;
+                ODataNestedResourceInfo navigationLink = new ODataNestedResourceInfo();
+                
+                //these items will all be of the same type ~ we'll be using the 
+                //first item in determining the type of the descriptor.
+                Descriptor item = relatedLinks.Value.FirstOrDefault();
+    
+                if (item.DescriptorKind == DescriptorKind.Entity)
+                {
+                    var entDescriptor = item as EntityDescriptor;
+
+                    if (entDescriptor.IsDeepInsert)
+                    {
+                        relatedEnd = entDescriptor.GetRelatedEnd();
+                    }
+                    else
+                    {
+                        relatedEnd = new LinkDescriptor(entDescriptor.ParentEntity, entDescriptor.ParentProperty, entDescriptor.Entity, this.requestInfo.Model);
+                    }
+                }
+                else
+                {
+                    relatedEnd = item as LinkDescriptor;
+                }
+
+                navigationLink.Name = relatedEnd.SourceProperty;
+                isCollection = clientType.GetProperty(relatedEnd.SourceProperty, UndeclaredPropertyBehavior.ThrowException).IsEntityCollection;
+
+                if (!started)
+                {
+                    odataWriter.WriteNestedResourceInfoStart(navigationLink);
+                    started = true;
+                }
+
+                odataWriter.WriteNestedResourceInfoStart(navigationLink, relatedEnd.Source, relatedEnd.Target);
+                if (isCollection)
+                {
+                    WriteFeedEntry(relatedLinks.Value, bulkUpdateGraph, odataWriter);
+                }
+                else
+                {
+                    //we can have a related single entry ~ this will automatically be a replace operation.
+                    if (item.DescriptorKind == DescriptorKind.Entity)
+                    {
+                        EntityDescriptor relatedEntintyDescriptor = item as EntityDescriptor;
+                        WriteDeltaEntry(relatedEntintyDescriptor, bulkUpdateGraph, odataWriter);
+                    }
+                    else if (item.DescriptorKind == DescriptorKind.Link)
+                    {
+                        odataWriter.WriteEntityReferenceLink(new ODataEntityReferenceLink() { Url = navigationLink.Url }, relatedEnd.Source, relatedEnd.Target);
+                    }
+                }
+
+                odataWriter.WriteNestedResourceInfoEnd(navigationLink, relatedEnd.Source, relatedEnd.Target);

```

Comment:
Two 'WriteNestedResourceInfoEnd' together looks incorrect?

Reply:
yes this is correct.